esmvalcore/preprocessor/_multimodel.py

```diff
@@ -248,115 +98,190 @@ def _unify_time_coordinates(cubes):
+    constraint = iris.Constraint(time=lambda cell: begin <= cell.point <= end)
+    return cube.extract(constraint)
+
+
+def _extend(cube, time_points):
+    """Extend cube to a specified time range.
+
+    If time points are missing before the start/after the end of the
+    time range, cubes for each missing time pointwith masked data will
+    be added to pad the time range to match `time_points`. This method
+    supports lazy operation.
+    """
+    cube.coord('time').bounds = None
+    cube_points = cube.coord('time').points
+
+    begin = cube_points[0]
+    end = cube_points[-1]
+
+    pad_begin = time_points[time_points < begin]
+    pad_end = time_points[time_points > end]
+
+    if (len(pad_begin)) == 0 and (len(pad_end) == 0):
+        return cube
+
+    template_cube = cube[:1].copy()
+    template_cube.data = np.ma.array(da.zeros_like(template_cube.data),
+                                     mask=True,
+                                     dtype=template_cube.data.dtype)
+
+    cube_list = []
+
+    for time_point in pad_begin:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
+
+    cube_list.append(cube)
+
+    for time_point in pad_end:
+        new_slice = template_cube.copy()
+        new_slice.coord('time').points = float(time_point)
+        cube_list.append(new_slice)
+
+    cube_list = iris.cube.CubeList(cube_list)
+
+    new_cube = cube_list.concatenate_cube()
+
+    return new_cube
+
+
+def _align(cubes, span):
+    """Expand or subset cubes so they share a common time span."""
+    _unify_time_coordinates(cubes)
 
+    if _time_coords_are_aligned(cubes):
+        return cubes
+
+    all_time_arrays = [cube.coord('time').points for cube in cubes]
 
-def _assemble_data(cubes, statistic, span='overlap'):
-    """Get statistical data in iris cubes."""
-    # New time array representing the union or intersection of all cubes
-    time_spans = [cube.coord('time').points for cube in cubes]
     if span == 'overlap':
-        new_times = reduce(np.intersect1d, time_spans)
+        common_time_points = reduce(np.intersect1d, all_time_arrays)
+        new_cubes = [_subset(cube, common_time_points) for cube in cubes]
     elif span == 'full':
-        new_times = reduce(np.union1d, time_spans)
-    n_times = len(new_times)
+        all_time_points = reduce(np.union1d, all_time_arrays)
+        new_cubes = [_extend(cube, all_time_points) for cube in cubes]
+    else:
+        raise ValueError(f"Invalid argument for span: {span!r}"
+                         "Must be one of 'overlap', 'full'.")
 
-    # Target array to populate with computed statistics
-    new_shape = [n_times] + list(cubes[0].shape[1:])
-    stats_data = np.ma.zeros(new_shape, dtype=np.dtype('float32'))
+    for cube in new_cubes:
+        # Make sure bounds exist and are consistent
+        cube.coord('time').bounds = None
+        cube.coord('time').guess_bounds()
```

Comment:
Maybe you can use the `get_time_bounds` function in `_time.py`. Sometimes the bounds that you get using `guess_bounds` are shifted or have negative values if the point starts at 0.

Reply:
Thanks for making me aware of this Is this an issue you are running into with the current implementation? Because in `master` we also use `guess_bounds` for the time coordinate. 

Looking at this, we would have to infer the frequency argument ourselves. Afaik it is not trivial to 'guess' the time frequency (it is not stored as an attribute on the cube), and it would have to work for all possible frequencies. I think the best way to go about it is to make the `frequency` argument optional and guess it in the `_get_time_bounds` function with `frequency=None`. 

I think that is a bit out of scope for this PR. And considering that this issue is present in the current `master` branch, could you make an issue so that we follow-up in a separate PR?

Reply:
Thanks @Peter9192 , good point, I overlooked that. I was looking at the `_align` function, where this may be more difficult.

Reply:
Just had a go at trying to see if there is an easy way to do this, but I do not see a way to do this cleanly. @valeriupredoi @sloosvel , would you be OK if we handle this in a separate PR or make it part of #744?

Reply:
either way works for me, cheers :+1:

Reply:
Sure, it's not that big of an issue anyway.